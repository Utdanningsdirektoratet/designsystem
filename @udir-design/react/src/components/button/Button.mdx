import { Meta, Canvas, Controls, Primary } from '@storybook/blocks';
import { Do, Dont, Stack } from '../../../.storybook/docs-components';
import { Button } from './Button';
import * as ButtonStories from './Button.stories';
import { ButtonEx1, ButtonEx2, ButtonEx4, ButtonEx5 } from './docs/doAndDont';

<Meta of={ButtonStories} />

# Button

Knapper lar brukerne utføre handlinger. Vi har variantene `primary`, `secondary` og `tertiary` som kommer i ulike fargevarianter og størrelser.

<Primary />
<Controls />

## Slik bruker du button

```tsx
import { Button } from '@udir-design/react';

<Button>Trykk på meg!</Button>;
```

## Varianter av button

Vi har forskjellige knapper for å fremheve utvalgte handlinger. `Primary`, `secondary` og `tertiary` bestemmer hvor viktig handlingen er.

### Primary button

Vi bruker denne knappen til den handlingen som er viktigst for brukeren, for eksempel Lagre, Neste eller Send inn. Det skal bare være en primærknapp per side.

<Canvas of={ButtonStories.Primary} />

### Secondary button

Vi bruker denne knappen til handlinger som ikke er hovedhandlinger. Den står gjerne sammen med en hovedknapp, så hvis hovedknappen er Lagre, kan sekundærknappen være Avbryt.

<Canvas of={ButtonStories.Secondary} />

### Tertiary button

Denne knappen bruker vi til nedtonede handlinger. Hvis vi vil bruke den alene, må den ha med et ikon, hvis ikke er det vanskelig å se at det er en knapp.

<Canvas of={ButtonStories.Tertiary} />

## Fargevarianter

Knappene kommer i to fargevarianter: `neutral` og `danger`.

- `neutral` er standardfargen.

<Canvas of={ButtonStories.Neutral} />

- `danger`-fargen bruker vi til handlinger som brukeren ikke kan angre, for eksempel Slett.

<Canvas of={ButtonStories.Danger} />

### Kombinere ulike fargevarianter

Vi kan bruke en `danger`-knapp sammen med én eller flere `neutral`-knapper når vi skal tydeliggjøre en handling brukeren ikke kan angre, for eksempel Forkast. Vi skal ikke kombinere andre fargevarianter.

<Canvas of={ButtonStories.CombinedColors} />

## Eksempler

### Knapp som lenke

Bruk `asChild` dersom du trenger å bruke knappen som en lenke. Vurder først om det er bedre å bruke [`Link`](/docs/components-link--docs)-komponenten.
Du kan lese mer om komposisjon [her](https://www.designsystemet.no/grunnleggende/for-utviklere/komposisjon).

<Canvas of={ButtonStories.AsLink} />

### Knapp med kun ikon

Knapper med kun ikon bruker vi bare for godt kjente ikoner som Lukk og Slett. Når knappen bare har ikon, må vi ha med en tilgjengelig tekst som forklarer hva knappen gjør. Dette kan legges i `title`, `aria-label` eller i en `Tooltip` som i eksempelet under.

<Canvas of={ButtonStories.IconOnly} />

### Knapp med tekst og ikon

Når vi bruker ikon sammen med tekst, skal ikonet ha `aria-hidden` slik at teksten ikke blir lest opp av skjermleser to ganger.

<Canvas of={ButtonStories.TextAndIcon} />

Vi plasserer ikonet til venstre eller høyre for teksten, avhengig av hva knappen kommuniserer og hvor den står i grensesnittet.

Som regel er det **venstre** som er den beste plasseringen, men det finnes unntak:

- Er det en knapp som sier “Start” med en høyrepekende pil, er det naturlig at pilen plasseres til høyre for teksten.
- Er knappen fast plassert på høyre side av grensesnittet er det best at ikonet også er høyrestilt for teksten.

### Størrelse på ikon

Ikon skalerer automatisk basert på knappestørrelse, men du kan endre denne ved å sette font størrelsen.

<Canvas of={ButtonStories.IconOnly} />

### Last inn

Når vi skal vise brukeren at knappen laster noe, kan vi kombinere knappen med en `spinner`.
I eksempelet under bruker vi proppen `loading` for å vise at knappen laster.

<Canvas of={ButtonStories.Loading} />

I eksempelet over blir `aria-disabled` satt av `loading="true"`. Dette er den anbefalte måten for å vise at en knapp er deaktivert. Da kan den fortsatt få fokus når noen navigerer med tastaturet. Slik vil blant annet skjermlesere og andre hjelpemidler bli informert om at knappen finnes, men ikke er aktiv.

Denne egenskapen stopper ikke automatisk knappen fra å utløse `onClick`.

Pass på at callback-funksjonen ikke kjører når knappen er deaktivert.

## Retningslinjer for button

Knapper skal brukes til å utføre handlinger.

**Passer til å**

- starte en prosess
- navigere i en flyt, for eksempel med Forrige og Neste
- sende inn et skjema

**Passer ikke til å**

- navigere til andre sider eller ut av tjenesten, bruk heller [`link`](/docs/components-link--docs)

### Plassering av knapper

Når vi legger flere knapper ved siden av hverandre, skal primærknappen stå først. Unntaket er Forrige og Neste, da skal sekundærknappen Forrige stå først.

### Ikke bruk deaktiverte knapper

Deaktiverte knapper bør unngås fordi de kan være vanskelige å oppfatte. Noen brukere vil ikke forstå hva knappen sier eller hvorfor den ikke er klikkbar. Bruk heller en aktiv knapp som validerer.

### En knapp skal utføre kun én handling

Vi skal tydelig beskrive handlingene knappene utfører. En knapp kan bare ha én handling, ikke flere.

<ButtonEx1 />

### Èn hovedhandling per side

Vi skal holde oss til én primærknapp per side. Hvis det er flere hovedhandlinger kan brukerne bli usikre på hva de skal gjøre.

<ButtonEx2 />

### Ett ikon per knapp

Vi skal holde oss til ett ikon per knapp.

<ButtonEx4 />

## Tekst i komponenten

Knappeteksten skal være kort, helst ikke mer enn to, maks. tre ord. Tenk på at knappen tydelig skal beskrive hva som skjer når brukeren klikker på den.

### Unngå å blande skrivemåter

Bruk stor forbokstav på det første ordet, som du ville gjort i en setning. Ikke bruk bare store bokstaver (all caps), det er vanskeligere å lese.

<ButtonEx5 />
