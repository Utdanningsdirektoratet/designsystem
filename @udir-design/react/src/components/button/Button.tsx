import {
  Button as DigdirButton,
  type ButtonProps as DigdirButtonProps,
} from '@digdir/designsystemet-react';
import { ForwardRefExoticComponent } from 'react';

type ButtonProps = Omit<DigdirButtonProps, 'data-color'> & {
  /**
   * Change the color scheme of the button
   */
  'data-color'?: 'neutral' | 'danger';
};

const Button = DigdirButton as ForwardRefExoticComponent<ButtonProps>;

export { Button, ButtonProps };
