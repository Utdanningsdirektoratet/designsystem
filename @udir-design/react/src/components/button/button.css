.ds-button {
  &[data-variant='tertiary']:not([data-color='danger']),
  &[data-variant='secondary']:not([data-color='danger']) {
    --dsc-button-color: var(--ds-color-text-default);
  }
}
