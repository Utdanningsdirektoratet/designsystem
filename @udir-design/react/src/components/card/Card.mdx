import { Meta, Canvas, Controls, Primary } from '@storybook/blocks';
import * as CardStories from './Card.stories';

<Meta of={CardStories} />

# Card

Vi bruker `Card` til å fremheve informasjon eller oppgaver som hører sammen. Kortene finnes i flere varianter og kan inneholde elementer som tekst, bilder, tekstfelt, knapper, lenker og videoer.

**Passer til å**

- gruppere innhold eller funksjonalitet som hører sammen
- fremheve innhold som er viktig for brukeren
- navigere til en annen side eller utføre en handling

**Passer ikke til å**

- erstatte seksjoner på en side (gruppere store mengder innhold i samme kort)

## Slik bruker du Card

<Primary />
<Controls />

```tsx
import { Card } from '@udir-design/react';
<Card>
  <Card.Block>Tittel</Card.Block>
  <Card.Block>Innhold</Card.Block>
  <Card.Block>Fotnote</Card.Block>
</Card>;
```

## Eksempler

### Med `Block`

Bruk flere `Card.Block` hvis du vil dele opp kortet med skillelinjer eller legge inn utfallende bilder eller video.
Merk at innhold kan ikke plasseres direkte i `Card` dersom du bruker `Card.Block`. Da burde alt innholdet være inni en av av kortets `Card.Block`-seksjoner.

<Canvas of={CardStories.Media} />

### Video

<Canvas of={CardStories.Video} />

### Lenkekort

`Card` kan brukes som lenkekort for å ta brukerne videre til en annen side.

**Legg inn lenken i kortets tittel.** Hele kortet blir da automatisk klikkbart, og skjermlesere får en [bedre brukeropplevelse enn dersom hele kortet var en lenke](https://adrianroselli.com/2020/02/block-links-cards-clickable-regions-etc.html).

<Canvas of={CardStories.WithLink} />

### Horisontalt

Bruk `display: flex` eller `display: grid` for å legge `Card.Block` ved siden av hverandre:

<Canvas of={CardStories.Horizontal} />

### Varianter og farger

Eksempler på `default`- og `tinted`-variantene med de forskjellige fargene.

<Canvas of={CardStories.Variants} />

### Komponert

Eksempel på hvordan andre komponenter kan brukes til å komponere et kort.

<Canvas of={CardStories.Composed} />

## Retningslinjer for Card

Følg disse retningslinjene for å sikre at kortene er enkle å bruke og forstå:

- Dersom kortet er en link med bilde, så skal bildet **ikke** være informativt, og `alt` attributtet skal være tomt.

- Et kort skal oppfattes som en selvstendig, avgrenset enhet. Innholdet i kortet skal være helhetlig og relatert til et enkelt emne eller en destinasjon.

- Kortet skal kunne stå alene uten å være avhengig av omkringliggende elementer for kontekst. Det skal gi tilstrekkelig informasjon til å forstå det presenterte innholdet uten eksterne referanser.

**Unngå å:**

- bruke for mange kort på en side, det kan skape forvirring
- blande for mange forskjellige typer innhold i et enkelt kort

## Tekst i komponenten

Unngå lange tekster på kortene. Tenk nøye gjennom hva du trenger å si og skriv en kort og presis overskrift, underoverskrift og eventuell brødtekst.
