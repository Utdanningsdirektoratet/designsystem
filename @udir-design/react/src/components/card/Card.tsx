import {
  Card,
  CardBlock,
  type CardBlockProps,
  type CardProps,
} from '@digdir/designsystemet-react';

// For some reason this fixes "ComponentSubcomponent" -> "Component.Subcomponent" in Storybook code snippets
Card.displayName = 'Card';

export { Card, CardBlock, CardBlockProps, CardProps };
