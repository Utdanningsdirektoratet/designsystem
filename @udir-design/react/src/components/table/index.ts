import {
  Table as DigdirTable,
  TableBody,
  type TableBodyProps,
  TableCell,
  type TableCellProps,
  TableFoot,
  type TableFootProps,
  TableHead,
  type TableHeadProps,
  TableHeaderCell,
  type TableHeaderCellProps,
  TableRow,
  type TableRowProps,
} from '@digdir/designsystemet-react';
import { Table as TableRoot, type TableProps } from './Table';

const Table = Object.assign(TableRoot, {
  Head: DigdirTable.Head,
  Body: DigdirTable.Body,
  Row: DigdirTable.Row,
  Cell: DigdirTable.Cell,
  HeaderCell: DigdirTable.HeaderCell,
  Foot: DigdirTable.Foot,
});

Table.displayName = 'Table';

export {
  Table,
  TableBody,
  TableBodyProps,
  TableCell,
  TableCellProps,
  TableFoot,
  TableFootProps,
  TableHead,
  TableHeadProps,
  TableHeaderCell,
  TableHeaderCellProps,
  TableProps,
  TableRow,
  TableRowProps,
};
