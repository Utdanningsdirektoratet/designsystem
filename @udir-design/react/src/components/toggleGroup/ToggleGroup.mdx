import { Meta, Canvas, Controls, Primary } from '@storybook/blocks';
import * as ToggleGroupStories from './ToggleGroup.stories';

<Meta of={ToggleGroupStories} />

# ToggleGroup

Med `ToggleGroup` kan brukerne velge alternativer som påvirker innholdet på en side. Komponenten består av en gruppe knapper som henger sammen, der kun én knapp er mulig å velge om gangen.

<Primary />
<Controls />

## Slik bruker du ToggleGroup

Plasser helst innholdet du skal styre rett etter `ToggleGroup` i DOM-en. Hvis ikke, bør du bruke `aria-controls` på `ToggleGroup` for å knytte komponenten til den containeren for innholdet som endrer seg.

```tsx
import { ToggleGroup } from '@udir-design/react';

<ToggleGroup defaultValue="value1">
  <ToggleGroup.Item value="value1">Option 1</ToggleGroup.Item>
  <ToggleGroup.Item value="value2">Option 2</ToggleGroup.Item>
  <ToggleGroup.Item value="value3">Option 3</ToggleGroup.Item>
</ToggleGroup>;
```

## Eksempler

### Innholdet på knappene

`ToggleGroup` kan ha tre typer innhold:

- Bare tekst.
- Tekst med ikon til venstre for teksten.
- Bare ikon.

#### Bare ikon

Alternativet "Bare ikon" brukes kun når du vet at brukerne vil forstå ikonet. Det kan for eksempel være ekspertbrukere eller brukere av interne tjenester, som kjenner symbolene i den aktuelle `ToggleGroup` godt. Forklar handlingene med [`Tooltip`](/docs/komponenter-tooltip--docs) når du bruker kun ikoner.

<Canvas of={ToggleGroupStories.OnlyIcons} />

### Kontrollert

Bruk `value` på `ToggleGroup.Item` og `ToggleGroup` for å kontrollere verdiene selv.

<Canvas of={ToggleGroupStories.Kontrollert} />

## Retningslinjer for ToggleGroup

Det bør være minst to, men ikke flere enn fem valg i en `ToggleGroup`. Trenger du flere valg enn fem, bruker du [`Combobox`](/docs/components-combobox--docs).

**Passer til å**

- filtrere innhold i lister og tabeller
- veksle mellom datavisning og innhold, for eksempel mellom tabell og graf
- styre hvilke felter i et skjema som skal vises

**Passer ikke til å**

- bruke som skjemaelement – bruk heller [`Radio`](/docs/components-radio--docs)
- navigere – bruk heller [`Tabs`](/docs/components-tabs--docs) hvis det dreier seg om panelnavigasjon

### Gi nok plass til komponenten

Pass på at hele komponenten får plass på skjermen og at alle valgene er godt synlig for brukerne. I grensesnitt med liten plass, for eksempel på mobil, kan vi bruke [`Combobox`](/docs/components-combobox--docs) i stedet.

## Tekst i komponenten

Ha gjerne med en ledetekst for gruppen, som forklarer hva valgene gjelder. Skriv tydelige tekster for valgene i komponenten, og pass på at de har samme språklige form. Ta med hint i [`Tooltip`](/docs/components-tooltip--docs), særlig hvis du bare viser ikoner.

## Tilgjengelighet

`ToggleGroup` har standard fokusindikator (ramme).
Vi bruker `roving tabindex` for å håndtere fokusbevegelse mellom knappene.

### Tastaturinteraksjon

- `Tab` Flytter fokuset til den aktive toggle-knappen.
- `Shift+Tab` Flytter fokuset til elementet før `ToggleGroup` i tabindex.
- `Mellomrom` Velger en knapp.
- `Enter` Velger en knapp.
- `Pil opp` Flytter fokuset til forrige knapp i gruppa.
- `Pil venstre` Flytter fokuset til forrige knapp i gruppa.
- `Pil ned` Flytter fokuset til neste knapp i gruppa.
- `Pil høyre` Flytter fokuset til neste knapp i gruppa.
- `Home` Går til første knapp i gruppa.
- `End` Går til siste knapp i gruppa.
