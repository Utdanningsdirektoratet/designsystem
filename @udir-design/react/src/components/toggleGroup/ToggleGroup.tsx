import {
  ToggleGroup,
  type ToggleGroupProps,
  ToggleGroupItem,
  type ToggleGroupItemProps,
} from '@digdir/designsystemet-react';

// For some reason this fixes "ComponentSubcomponent" -> "Component.Subcomponent" in Storybook code snippets
ToggleGroup.displayName = 'ToggleGroup';

export { ToggleGroup, ToggleGroupProps, ToggleGroupItem, ToggleGroupItemProps };
