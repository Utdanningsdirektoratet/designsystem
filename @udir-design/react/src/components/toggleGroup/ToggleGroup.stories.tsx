import type { Meta, StoryObj } from '@storybook/react-vite';
import { useState } from 'react';
import { expect, fn, userEvent, within } from 'storybook/test';
import {
  AlignCenterIcon,
  AlignLeftIcon,
  AlignRightIcon,
  ArrowRightIcon,
  CheckmarkCircleIcon,
  DocPencilIcon,
  ExclamationmarkTriangleIcon,
  NotePencilDashIcon,
  NotePencilIcon,
  TasklistIcon,
  XMarkOctagonIcon,
} from '@udir-design/icons';
import { demoChromaticModes } from 'src/demo/demoParameters';
import { Button } from '../button/Button';
import { Card } from '../card/Card';
import { Table } from '../table';
import { Tooltip } from '../tooltip/Tooltip';
import { Heading } from '../typography/heading/Heading';
import { ToggleGroup } from './ToggleGroup';

const meta: Meta<typeof ToggleGroup> = {
  component: ToggleGroup,
  tags: ['beta', 'digdir'],
  parameters: {
    componentOrigin: {
      originator: 'digdir',
      details: 'Vi har fjernet mulighet for fargevalg.',
    },
    layout: 'centered',
  },
};

export default meta;
type Story = StoryObj<typeof ToggleGroup>;

export const Preview: Story = {
  args: {
    defaultValue: 'innboks',
    onChange: fn(),
  },
  render: (args) => (
    <ToggleGroup {...args}>
      <ToggleGroup.Item value="innboks">Innboks</ToggleGroup.Item>
      <ToggleGroup.Item value="utkast">Utkast</ToggleGroup.Item>
      <ToggleGroup.Item value="arkiv">Arkiv</ToggleGroup.Item>
      <ToggleGroup.Item value="sendt">Sendt</ToggleGroup.Item>
    </ToggleGroup>
  ),
  play: async ({ canvasElement, step, args }) => {
    const canvas = within(canvasElement);
    const innboksButton = canvas.getByRole('radio', { name: /innboks/i });
    const utkastButton = canvas.getByRole('radio', { name: /utkast/i });
    const arkivButton = canvas.getByRole('radio', { name: /arkiv/i });

    await step('Default selection is "Innboks"', async () => {
      expect(innboksButton).toHaveAttribute('aria-checked', 'true');
    });

    await step('Only one ToggleGroup item is active initially', async () => {
      const buttons = canvas.getAllByRole('radio');
      const activeButtons = buttons.filter(
        (btn) => btn.getAttribute('aria-checked') === 'true',
      );
      expect(activeButtons).toHaveLength(1);
    });

    await step('User can navigate with arrow keys', async () => {
      await userEvent.tab();
      expect(innboksButton).toHaveFocus();

      await userEvent.keyboard('{arrowright}');
      expect(utkastButton).toHaveFocus();

      await userEvent.keyboard('{arrowright}');
      expect(arkivButton).toHaveFocus();

      await userEvent.keyboard('{arrowleft}');
      expect(utkastButton).toHaveFocus();
    });

    await step(
      'Clicking a different option updates the active state and calls onChange',
      async () => {
        expect(utkastButton).toHaveAttribute('aria-checked', 'false');
        await userEvent.click(utkastButton);
        expect(args.onChange).toHaveBeenCalledWith('utkast');
        expect(utkastButton).toHaveAttribute('aria-checked', 'true');
        expect(innboksButton).toHaveAttribute('aria-checked', 'false');
      },
    );

    await step(
      'Clicking an already active option keeps it active',
      async () => {
        expect(utkastButton).toHaveAttribute('aria-checked', 'true');
        await userEvent.click(utkastButton);
        expect(utkastButton).toHaveAttribute('aria-checked', 'true');
      },
    );
  },
};

export const Primary: Story = {
  args: { defaultValue: 'innboks', variant: 'primary' },
  render: Preview.render,
};

export const Secondary: Story = {
  args: { defaultValue: 'innboks', variant: 'secondary' },
  render: Preview.render,
};

export const OnlyText: Story = {
  args: {},
  parameters: {
    customStyles: {
      display: 'flex',
      flexDirection: 'column',
      gap: 'var(--ds-size-2)',
    },
  },
  render: () => (
    <>
      <label>Filtrering av skjema</label>
      <ToggleGroup defaultValue="personlig">
        <ToggleGroup.Item value="personlig">Personlig</ToggleGroup.Item>
        <ToggleGroup.Item value="generelt">Generelt</ToggleGroup.Item>
        <ToggleGroup.Item value="tilleggsinformasjon">
          Tilleggsinformasjon
        </ToggleGroup.Item>
      </ToggleGroup>
    </>
  ),
};

export const TextAndIcons: Story = {
  args: {
    defaultValue: 'option-1',
  },
  render: () => (
    <ToggleGroup defaultValue="aktiv">
      <ToggleGroup.Item value="aktiv">
        <CheckmarkCircleIcon aria-hidden />
        Aktiv
      </ToggleGroup.Item>
      <ToggleGroup.Item value="advarsel">
        <ExclamationmarkTriangleIcon aria-hidden />
        Advarsel
      </ToggleGroup.Item>
      <ToggleGroup.Item value="kritisk">
        <XMarkOctagonIcon aria-hidden />
        Kritisk
      </ToggleGroup.Item>
    </ToggleGroup>
  ),
};

export const OnlyIcons: Story = {
  args: {
    defaultValue: 'venstrestilt',
  },
  render: (args) => (
    <ToggleGroup {...args}>
      <Tooltip content="Venstrestilt">
        <ToggleGroup.Item value="venstrestilt" icon aria-label="Venstrestilt">
          <AlignLeftIcon aria-hidden />
        </ToggleGroup.Item>
      </Tooltip>
      <Tooltip content="Midtstilt">
        <ToggleGroup.Item value="midtstilt" icon aria-label="Midtstilt">
          <AlignCenterIcon aria-hidden />
        </ToggleGroup.Item>
      </Tooltip>
      <Tooltip content="Høyrestilt">
        <ToggleGroup.Item value="høyrestilt" icon aria-label="Høyrestilt">
          <AlignRightIcon aria-hidden />
        </ToggleGroup.Item>
      </Tooltip>
    </ToggleGroup>
  ),
};

const testsData: Record<
  string,
  Array<{ name: string; subject: string; grade?: string; delivered?: Date }>
> = {
  answers: [
    {
      name: 'Ola Nordmann',
      subject: 'Matematikk',
      delivered: new Date('2023-10-01'),
    },
    {
      name: 'Kari Nordmann',
      subject: 'Engelsk',
      delivered: new Date('2023-10-02'),
    },
    {
      name: 'Per Hansen',
      subject: 'Naturfag',
      delivered: new Date('2023-10-03'),
    },
    {
      name: 'Stian Berg',
      subject: 'Norsk',
      delivered: new Date('2023-10-04'),
    },
  ],
  drafts: [
    {
      name: 'Stian Berg',
      subject: 'Matematikk',
      grade: '4',
    },
    {
      name: 'Kari Nordmann',
      subject: 'Norsk',
      grade: '5',
    },
    {
      name: 'Per Hansen',
      subject: 'Norsk',
      grade: '4',
    },
  ],
  correctedAnswers: [
    {
      name: 'Per Hansen',
      subject: 'Naturfag',
      grade: '6',
    },

    { name: 'Ola Nordmann', subject: 'Engelsk', grade: '4' },
    { name: 'Kari Nordmann', subject: 'Engelsk', grade: '5' },
  ],
};

export const Controlled: Story = {
  parameters: {
    chromatic: {
      modes: {
        minimum: demoChromaticModes.minimum,
        desktop: demoChromaticModes.desktop,
      },
    },
    customStyles: {
      display: 'flex',
      flexDirection: 'column',
      gap: '1rem',
      maxWidth: '750px',
    },
  },
  render: function Render(args) {
    const [value, setValue] = useState<string>('correctedAnswers');
    const isAnswers = value === 'answers';
    const actionData = {
      answers: { text: 'Start retting', icon: <NotePencilIcon aria-hidden /> },
      drafts: {
        text: 'Rediger utkast',
        icon: <NotePencilDashIcon aria-hidden />,
      },
      correctedAnswers: {
        text: 'Se retting',
        icon: <ArrowRightIcon aria-hidden />,
      },
    } as const;
    const actionText = (currentValue: string): string =>
      actionData[currentValue as keyof typeof actionData]?.text || '';
    const actionValue = (currentValue: string) =>
      actionData[currentValue as keyof typeof actionData]?.icon || null;
    return (
      <>
        <Heading level={1}>Retting av prøver</Heading>
        <ToggleGroup value={value} onChange={setValue} {...args}>
          <ToggleGroup.Item value="answers">
            <TasklistIcon aria-hidden />
            Ikke påbegynt
          </ToggleGroup.Item>
          <ToggleGroup.Item value="drafts">
            <DocPencilIcon aria-hidden />
            Under arbeid
          </ToggleGroup.Item>
          <ToggleGroup.Item value="correctedAnswers">
            <CheckmarkCircleIcon aria-hidden />
            Ferdig
          </ToggleGroup.Item>
        </ToggleGroup>

        <Table style={{ tableLayout: 'fixed' }}>
          <Table.Head>
            <Table.Row>
              <Table.HeaderCell>Navn</Table.HeaderCell>
              <Table.HeaderCell>Fag</Table.HeaderCell>
              <Table.HeaderCell>
                {isAnswers ? 'Innlevert' : 'Karakter'}
              </Table.HeaderCell>
              <Table.HeaderCell style={{ textAlign: 'right' }}>
                {actionText(value)}
              </Table.HeaderCell>
            </Table.Row>
          </Table.Head>
          <Table.Body>
            {testsData[value].map((item) => (
              <Table.Row key={item.name}>
                <Table.Cell>{item.name}</Table.Cell>
                <Table.Cell>{item.subject}</Table.Cell>
                <Table.Cell>
                  {isAnswers
                    ? item.delivered?.toLocaleDateString()
                    : item.grade}
                </Table.Cell>
                <Table.Cell
                  style={{
                    display: 'flex',
                    justifyContent: 'flex-end',
                  }}
                >
                  <Tooltip content={actionText(value)}>
                    <Button
                      variant="tertiary"
                      icon
                      aria-label={actionText(value)}
                    >
                      {actionValue(value)}
                    </Button>
                  </Tooltip>
                </Table.Cell>
              </Table.Row>
            ))}
          </Table.Body>
        </Table>
        {!isAnswers && (
          <Button
            style={{ width: 'fit-content', marginTop: 'var(--ds-size-4)' }}
            variant="secondary"
            onClick={() => setValue('answers')}
          >
            Se ikke påbegynt
          </Button>
        )}
      </>
    );
  },
};

export const ToggleGroupInColorContext: Story = {
  args: Preview.args,
  render: (args, context) => (
    <Card data-color="accent" variant="tinted">
      {Preview.render?.(args, context)}
    </Card>
  ),
  play: async ({ canvasElement, step }) => {
    await step(
      'Should have neutral color palette by default, no matter the surrounding color palette',
      async () => {
        const firstButton = within(canvasElement).getAllByRole('radio')[0];
        const expectedColor = getComputedStyle(firstButton).getPropertyValue(
          '--ds-color-neutral-base-default',
        );
        await expect(firstButton).toHaveStyle(
          `background-color: ${expectedColor}`,
        );
      },
    );
  },
};
