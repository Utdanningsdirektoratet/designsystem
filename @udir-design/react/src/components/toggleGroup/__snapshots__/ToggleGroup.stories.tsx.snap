// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Preview 1`] = `
<div
  role="radiogroup"
  tabindex="0"
>
  <button
    data-variant="primary"
    type="button"
    value="innboks"
    id="togglegroup-item-:r2:"
    aria-checked="true"
    aria-current="true"
    role="radio"
    name="togglegroup-name-:r0:"
    data-roving-tabindex-item="true"
    tabindex="0"
  >
    Innboks
  </button>
  <button
    data-variant="tertiary"
    type="button"
    value="utkast"
    id="togglegroup-item-:r4:"
    aria-checked="false"
    aria-current="false"
    role="radio"
    name="togglegroup-name-:r0:"
    data-roving-tabindex-item="true"
    tabindex="-1"
  >
    Utkast
  </button>
  <button
    data-variant="tertiary"
    type="button"
    value="arkiv"
    id="togglegroup-item-:r6:"
    aria-checked="false"
    aria-current="false"
    role="radio"
    name="togglegroup-name-:r0:"
    data-roving-tabindex-item="true"
    tabindex="-1"
  >
    Arkiv
  </button>
  <button
    data-variant="tertiary"
    type="button"
    value="sendt"
    id="togglegroup-item-:r8:"
    aria-checked="false"
    aria-current="false"
    role="radio"
    name="togglegroup-name-:r0:"
    data-roving-tabindex-item="true"
    tabindex="-1"
  >
    Sendt
  </button>
</div>
`;

exports[`Toggle Group In Color Context 1`] = `
<div
  data-variant="tinted"
  data-color="accent"
>
  <div
    role="radiogroup"
    tabindex="0"
  >
    <button
      data-variant="primary"
      type="button"
      value="innboks"
      id="togglegroup-item-:rb:"
      aria-checked="true"
      aria-current="true"
      role="radio"
      name="togglegroup-name-:r9:"
      data-roving-tabindex-item="true"
      tabindex="0"
    >
      Innboks
    </button>
    <button
      data-variant="tertiary"
      type="button"
      value="utkast"
      id="togglegroup-item-:rd:"
      aria-checked="false"
      aria-current="false"
      role="radio"
      name="togglegroup-name-:r9:"
      data-roving-tabindex-item="true"
      tabindex="-1"
    >
      Utkast
    </button>
    <button
      data-variant="tertiary"
      type="button"
      value="arkiv"
      id="togglegroup-item-:rf:"
      aria-checked="false"
      aria-current="false"
      role="radio"
      name="togglegroup-name-:r9:"
      data-roving-tabindex-item="true"
      tabindex="-1"
    >
      Arkiv
    </button>
    <button
      data-variant="tertiary"
      type="button"
      value="sendt"
      id="togglegroup-item-:rh:"
      aria-checked="false"
      aria-current="false"
      role="radio"
      name="togglegroup-name-:r9:"
      data-roving-tabindex-item="true"
      tabindex="-1"
    >
      Sendt
    </button>
  </div>
</div>
`;
